### ISSUE TITLE
Option to display data as QR code

### ISSUE THREAD

**Issue by sowbug** [keepassxreboot/keepassxc#675]
Sometimes it's useful to transfer secrets via QR code, and it would be nice if KeePassXC offered that feature. It could be another button icon next to the eye icon that shows a passphrase.

Use cases
- transferring a strong passphrase to a mobile phone (hard to type because it's mobile and because the passphrase is strong).
- transferring information from an offline instance of the application, where you don't want to risk inserting a USB drive.

**Comment by phoerious** [comment]
That would require a safe scanner app on the phone, though, i.e., one that doesn't store the plaintext password in its history.

**Comment by sowbug** [comment]
Yes, that's right. The way I look at it, the more you facilitate using very strong passphrases, the more people will be willing to use strong passphrases. And the easier you make it to export specific passphrases to not-necessarily-safe environments such as a mobile phone, the more likely you are to be willing to adopt an overall safer passphrase store, such as an offline and/or completely trusted machine running a KeePass variant.

If it's a pain in the neck to get passphrases out without connecting safe and unsafe devices, or laboriously retyping 30-character passphrases, you'll tend to use weaker passphrases on the unsafe devices.

**Comment by phoerious** [comment]
I feel your pain. It gets especially hard when your password contains non-ASCII characters. We just need to make sure, users don't shoot themselves in the foot with this.

**Comment by TheZ3ro** [comment]
It has been already discussed here #96

**Comment by adolfogc** [comment]
I was thinking on this lately. What do you think of developing a minimal mobile app for this using Qt? Maybe we could use _callback-urls_ (URL schemes) and encrypt the passwords, so that it doesn't matter which app is used when reading the QR code. For instance, iOS 11 users could read the QR code using the native camera app and android users could read them using the Barcode Scanner app. The keepassxc mobile app will then decrypt the password and administer the password.

**Comment by VinArt** [comment]
A friend of mine was asking for this feature, to display a password as QR code because of exactly having to type a strong password on mobile. So I implemented in entry edit widget a button that toggles a pixmap (200x200px) with a QR of plain-text password that can be scanned by any reader. Does anyone want this feature to be here?

**Comment by adolfogc** [comment]
I'm curious about how you encoded the password into a QR code. Which library did you use for this project? Please see the approach I'm trying in PR #1001 for implementing something related using Libqrencode.

Also, can you share a screenshot of the interface you're describing? 

Another use-case I was thinking about is displaying QR codes for WiFi passwords using [this format](https://github.com/zxing/zxing/wiki/Barcode-Contents#wifi-network-config-android) that is supported by both Android and iOS11. This is useful for connecting mobile devices quickly to WiFi, especially when you use complicated/long passwords. What do you think is a good way to implement this in the UI?

One potential issue I see is the QR encoding library leaking information through freed memory. Initially, I was thinking on zeroing the buffer passed to Libqrencode using C11's `memset_s` (see [this link](http://en.cppreference.com/w/c/string/byte/memset) for more info.), but came across two potential problems:
- I'm not sure C11's `memset_s` is implemented in the all the possible target platforms;
- a quick inspection of libqrencode's source gave me the impression that it is copying the data we give it to other internal buffers that we can't zero from our C++ code.

**Comment by adolfogc** [comment]
Another thought: maybe we can ask some of the most popular open source KeePass mobile implementations, e.g. MiniKeePass, to support the Android/iOS part of this?

**Comment by affinityv** [comment]
I've used QrCodeGenerator.plgx [1] in Winblows.

Today I use qrencode from Linux CLI, but it is messier.
For example (on Linux):
qrencode -t utf8 'test password'

It would definitely be better to be able to do this the KP2 (using that plugin).
[1] https://keepass.info/plugins.html#qrcodegen

**Comment by affinityv** [comment]
The plugin also works with KP2 (under mono) with Linux.  Right click on an entry and display the QR code.

**Comment by VinArt** [comment]
So my UI looks like this, when you open the entry:
![qr_empty](https://user-images.githubusercontent.com/3738015/32317723-16380ae4-bfb5-11e7-8107-368ce6cc8756.png)
and when you hit Show/hide QR like this:
![qr_full](https://user-images.githubusercontent.com/3738015/32317725-16592d3c-bfb5-11e7-8455-ab44391432e7.png)
I used a tiny QR implementation from https://github.com/mrdeveloperdude/OctoMY/tree/master/src/libs/libqr/qr which instead is a version of this https://github.com/nayuki/QR-Code-generator/tree/master/cpp or maybe even exactly that. 

I have not though of the security point of view yet, this is just to show how it can look like, but I think that after generation of Qr it is possible to zerofill memory.

**Comment by affinityv** [comment]
It would be better to have the QR code outside in it's own "window" so that it can be resized easily (sometimes it is useful to resize QR codes, especially if they are complex).

**Comment by VinArt** [comment]
I think I could do right click-> show QR and make a modal popup with QR code. Allthough I tested with up to 128 characters [0-9a-zA-Z/*...]+extended ASCII and with 200x200 I had no problem scanning the code.
![image](https://user-images.githubusercontent.com/3738015/32340229-483218ba-bffa-11e7-9c0a-3ac1762c2b17.png)
But yes, having a context-menu entry to show QR makes total sense. Especially with a hot-key and closing with ESC.

I will work further on this (also security-wise) and post after having some progress.

**Comment by adolfogc** [comment]
@VinArt Thanks for bringing Nayuki's QR code library to my attention, it seems like a nice alternative to use. I'll try to update my PR to use this library instead. Regarding the UI, I don't like how the empty white box looks and agree with you and @affinityv on using a modal window as an alternative.

**Comment by adolfogc** [comment]
@VinArt Please review PR #1167. I think we can group all QR code functionality under the same option: `WITH_XC_QRCODE`.

**Comment by phoerious** [comment]
Do we really need yet another compile-time parameter?

**Comment by adolfogc** [comment]
@phoerious Do you mean QR code functionality should be enabled by default?

**Comment by phoerious** [comment]
I don't see why not. Yes, technically it's another dependency, but nothing big.

**Comment by adolfogc** [comment]
Ok, I'll modify PR #1167 to reflect this change.

**Comment by adolfogc** [comment]
@phoerious I've removed the compile-time flag as per your request.

**Comment by VinArt** [comment]
@adolfogc I was not on this for a while, did you implement QR code display for all passwords as discussed or only for one-time? Is there something I could/should do?

**Comment by adolfogc** [comment]
@VinArt Sorry for my late reply as well; this has been a busy week at work. 

> ...did you implement QR code display for all passwords as discussed or only for one-time?

So far, I've adapted Project Yakuyi's `qrcodegen` for use with KeePassXC and implemented the functionality for the TOTP part.

> Is there something I could/should do?

Sure. The remaining functionality we've discussed in this thread remains to be implemented. IMHO, further discussion remains to be had in order to reach consensus on which will be the way to implement the part of displaying a password as a QR code to enable transfering it to a mobile device, so I think a next step could be implementing the part of displaying WiFi passwords in the special format supported by both iOS and Android. This could be done using the URL field of an entry and the placeholder funcionality of KeePassXC, for example, in the URL field we could write:

`WIFI:T:WPA;S:mynetwork;P:{PASSWORD};;`

Then, we can just look at an entry's URL field to check if it starts with "WIFI:" and, if so, show a little QR code button next to it that, when clicked, shows a modal with the URI rendered as a QR code. What do you think of this approach? And, would you like to work on this?

Before working on this, though, we should wait to hear what others think of implementing this new feature.

**Comment by VinArt** [comment]
@adolfogc No problem =) 
I would like to work on this. A couple of remarks:
1. Putting template for QR in URL field might be not very nice, since there is a possibility that that field will be in use already.
2. Maybe we should think of fields in settings where the user could input their templates. Maybe someone would like to QR just {PASSWORD}. Here we can provide a number of pre-defined templates, for example.
3. In the entry we can create a checkbox (enable QR) with a dropdown to select a template to use with this entry.

**Comment by adolfogc** [comment]
@VinArt 

> Putting template for QR in URL field might be not very nice, since there is a possibility that that field will be in use already.

You're right. I was thinking merely on entries for WiFi passwords. I guess that using {PASSWORD} as the sole placeholder would provide a means for users to accomplish what was originally requested on this issue. It would also require explicit intent from the user; using it that way would be sort of a "at your own risk" feature for power users.

> Maybe we should think of fields in settings where the user could input their templates. Maybe someone would like to QR just {PASSWORD}. Here we can provide a number of pre-defined templates, for example.


> In the entry we can create a checkbox (enable QR) with a dropdown to select a template to use with this entry.

Instead of creating new settings and using a checkbox, would it be better to just define a new attribute for each entry and use the placehoders there? We could agree on a convention to use. For instance, looking for an attribute using the special name of `qrcode`. For each entry, the QR button would be  shown only if such an attribute exists.

For the button, I was thinking o a representative QR code icon, like the ones shown here: https://tagliala.github.io/vectoriconsroundup/#qrcode

**Comment by TheZ3ro** [comment]
What about following the TOTP method?
When right-clicking an entry you have the `QR Code` menu where you can Show the QR or setup QR.

Setup QR will be a dialog where user can set a placeholder string or whatever he wants to display in the QR (like `WIFI:T:WPA;S:mynetwork;P:{PASSWORD};;` or `{S:WiFi}`) and then save this information in a additional attribute named `QR_string`

Show QR will be a dialog displaying the QR_string

**Comment by adolfogc** [comment]
@TheZ3ro I like your proposal. It makes this feature more usable for all users in general (not just power users) while at the same time keeping it from being a "hidden feature".

**Comment by TheZ3ro** [comment]
I think we should drop the WEP one 😅

Anyway yes, a preset for WiFi-format password, password only (`{PASSWORD}`) and username-password ({USERNAME}{TAB}{PASSWORD}). This string will also be editable so advanced users can insert whatever they want

**Comment by adolfogc** [comment]
> I think we should drop the WEP one 😅

Yes. 😅 I thought it may be needed in case someone had a legacy device, but now that you pointed out, I guess this won't be the case for any KeePassXC user.

**Comment by XIncognito10** [comment]
any update on this?

**Comment by Ceaus** [comment]
The "safe/secure QR scanner on your phone"  argument is a silly argument. I have hundreds of passwords in my database. Of which 98% are non-important (anyone wants my Kawasaki Versys forum password?). I also don't want the complicated route of emailing passwords to myself. So generating a QR code on my laptop from withing KPXC and than scanning the QR code on my phone is by far the most practical approach of transferring passwords from computer to phone.
BTW, I have also other bits in my KP database, for records keeping which are not security related, such as my full bank account number. With 20+ characters that is not something easy to remember and/or type when someone asks me for if by chat/txt/sms.
This issue is not a security issue. Not by a stretch.

**Comment by phoerious** [comment]
It's not silly. You don't want to scan your passwords and then realise a month later that all of them were saved enencryptedly in your scanner's history.

**Comment by Ceaus** [comment]
So you deny me (and thousands of other people) easy access to 98% of my non-important passwords because you think I need to be protected from myself for the remaining 2%? On my encrypted phone? Really? I can still think and decide for my self how I manage those really important passwords.

I'd say the risk associated with QRing passwords is infinitesimally smaller than forcing people to go back to "Welcome01"  because an app thinks it knows better.

**Comment by phoerious** [comment]
You may be able to "think and decide for yourself", but others may not be as informed about the risks as you are. Your forum password may not be very important, but someone else may have a password and if it got leaked, somebody could actually die.

**Comment by phoerious** [comment]
I have had a lot of discussions with security trainers of high-risk user groups such as human rights activists in countries without free speech. I think I an assess those issues a lot better than you think.

**Comment by phoerious** [comment]
So, don't get me wrong. I'm not against having this feature. I absolutely see its usefulness and would use it myself. I'm only against adding it to KeePassXC without discussing possible threat scenarios first and finding solutions to them in order to keep our users safe. KeePassXC has become so much more than just a tool made by tech-savvy geeks for tech-savvy geeks.

**Comment by Ceaus** [comment]
> I have had a lot of discussions with security trainers of high-risk user groups such as human rights activists in countries without free speech. I think I can assess those issues a lot better than you think.

And that is why your reasoning is flawed. There is no chance (none) that a person in a high risk environment, coached by trainers would use a crappy Android phone to store password if their lives would depend on it. Please, find me one (only ONE) who would trust a life or death password to a phone.
If you have found that person, please confirm that person would QR that password so that he can enter it easily from his phone into a website.
I don't get it, why security discussions like this always end up in some fictive, preposterous life and death scenario. It's almost Godwin's law. It is so totally unbelievable. I don't even know where to begin. It just doesn't make sense, how hard you may try. 

And now you also understand you're confusing security with safety. Don't pretend KPXC is about safety. It's not.

>  without discussing possible threat scenarios first

The only threat scenarios can pertain to KPXC itself (database consistency, entropy, hash algorithms, etc). Everything outside of KPXC is outside your realm of influence.

And this is the last I will say about it. Please understand the difference between safety and security, and how KPXC plays it's role in this, before bring up the subject again.

**Comment by phoerious** [comment]
> Please, find me one (only ONE) who would trust a life or death password to a phone.

Oh, I could probably find you thousands. Your conception is way off here. What the trainers try to do is make people aware of possible dangers in basic training sessions of a few days, sometimes up to a week or so. Don't confuse it with full education in digital security.

> And now you also understand you're confusing security with safety. Don't pretend KPXC is about safety. It's not.

Yes, it is.

> And this is the last I will say about it. Please understand the difference between safety and security, and how KPXC plays it's role in this, before bring up the subject again.

Yeah, I'm just the maintainer of KeePassXC. How would I know anything about this project?

**Comment by mt7479** [comment]
I think the feature could be included like the legacy browser extension.

* Disabled by default
* Checkbox to enable qr code functionality
* Display a warning text:

It is very likely that scanned passwords will be stored un-encrypted on he target device. Use witch caution for high security passwords.

* Possibly a link to website going into more detail.

This should be suficient to warn an uneducated user and provide the feature for those people that want to use it.

**Comment by scorcher24** [comment]
I want to add my voice to this. I was using a QR plugin for the plain old KeePass to transfer my long passwords to my phone, since I do not want the database on it.  Doing it this way makes the whole process a lot less painful. I am aware of the risks.

**Comment by fireglow** [comment]
Please add this.
From what I've understood the reason this features faces criticism is that a 3rd party app POSSIBLY COULD store the password in plain text.
That is entirely out of your control.
It's the same as saying "Some 3rd party app could access the computers memory and read sensitive data directly from KeePassXC". It's absurd.
What you an do is say "We recommend you scan passwords with app XYZ.
Because this feature is not implemented because of this faulty line of reasoning, users like myself resort to calling `qrencode` from the command line, so the password is in the history file. Hardly an improvement.
Focus on the things that you can do to improve data leakage in your own project. Offering a way to export data without extra steps or tools (WHICH WILL LEAK DATA) seems like a good idea.

**Comment by droidmonkey** [comment]
Thank you for your feedback.

**Comment by PaddeK** [comment]
I would like this QR feature too. Like @fireglow i also think the reasoning against this feature is faulty. Right now i could copy&paste a password and put it in a plain text file. You simply cant prevent users from doing stupid/unsecure things. Informing the user about the risks, making the feature optional and implementing it so you could show/hide the QR code is all you can do and frankly all that's necessary in my opinion.

**Comment by IPv777** [comment]
+1 for the feature request :)
I use https://f-droid.org/fr/packages/de.markusfisch.android.binaryeye/ and this feature would be very useful to me :D

**Comment by Insoleet** [comment]
I'd love this feature, so that I could migrate from Keepass to Keepass XC.

**Comment by almariah** [comment]
Actually it a very good feature, I need this for sharing a strong password with my mobile/tablets, etc. I have wrote something for Linux/macOS which is very basic to have another tool for me to add this feature

https://github.com/pharmatics/qr-secret

Also, keep in mind that installing keepassxc is easier than keepass (with QR plugin) in macOS and other distros like CentOS.

**Comment by almariah** [comment]
For scanning part, the Kaspersky QR scanner has a feature of disabling scanning history

https://www.kaspersky.de/qr-scanner

**Comment by iromeister** [comment]
As I've started using a smartphone only lately, I'm also late to this discussion... But I'd really like to have such a feature for the few occasions where for now I have to type in 20 random characters on my smartphone.

**Comment by moerkey** [comment]
+1 for this feature. It is not your responsibility to choose which app the user should use to read the QR code securely.

**Comment by noraj** [comment]
Some OTP client mobile app have an embedded QR code scanner without history & all of course so maybe some android implementation of keepass also have some embedded QR code scanner too.
Anyway that's not the point, it's up to the user to evaluate it's own risk because it seems less risky to swan a qr code with a local QR code app even if it store it than copy/paste it on an email to yourself and syncing it on a third party cloud forever + it is also stored on the email client app unless too you remove it manually as you can remove the qr code scanner history manually too.

**Comment by jvalrog** [comment]
+1 
I absolutely detest having to manually type a 20+ random characters password on my phone/tablet from the keepassxc app on desktop.

QR would be awesome. If bitcoin wallets have it, why can't this one?

**Comment by droidmonkey** [comment]
KeePassDX and KeePass2Android are two excellent apps that open kdbx databases and are fully compatible with KeePassXC. There is no reason not to use one of these apps. Transfer your database with your favorite cloud storage or manually with network shares/USB device.

**Comment by moerkey** [comment]
@droidmonkey On which data are you referring on when you speak of "vast majority"?

Can we pay you to get the feature in a easter egg way and useable only by a special flag? ;)

**Comment by Ceaus** [comment]
> Regarding users evaluating their own risk... the vast majority of users have no idea what the risk is. Offering the option allows them to make bad decisions and expose information unknowingly. That is the exact opposite of our application's goals.

We've been through this so many times by now. You don't know what my risk profile is. And you also don't know if my decision is a bad one. Please stop this line of reasoning. If you insist in knowing it better, then simply put a big warning box next to the QR code to tell them it's not safe.

**Comment by droidmonkey** [comment]
Just going to leave this here https://arstechnica.com/information-technology/2021/02/android-barcode-scanner-with-10-million-downloads-infects-users/

**Comment by phoerious** [comment]
Reminds me of the Stylish browser extension which suddenly went from useful app to full spyware with an ostensibly innocuous update.

**Comment by rocketraman** [comment]
> Just going to leave this here https://arstechnica.com/information-technology/2021/02/android-barcode-scanner-with-10-million-downloads-infects-users/

At the end of the day keepassXC could also suddenly include malware (either the developers go rogue :-) or more likely a scenario like Solarwinds that affects the distribution chain) and send all my passwords to a malicious actor -- a *far* more damaging scenario then this. The barcode scanner, as far as I can tell from the article, did not send the content of the stuff it was scanning out into the wild -- it just set up adware on the user's browser. Not great, but not the end of the world either.

If I was that paranoid, I wouldn't trust KeepassXC either.

**Comment by droidmonkey** [comment]
Just caught my attention and related to the discussion, that is all

**Comment by Ceaus** [comment]
Op 08-02-2021 om 22:43 schreef Jonathan White:
>
> Just going to leave this here 
> https://arstechnica.com/information-technology/2021/02/android-barcode-scanner-with-10-million-downloads-infects-users/ 
> <https://arstechnica.com/information-technology/2021/02/android-barcode-scanner-with-10-million-downloads-infects-users/>
>

I think this news is distorting the discussion of this github issue. 
There will always be rogue applications and libraries.

To me, in this particular case, the issue is why the KeepassXC devs are 
on their high horse to tell us what's good for us. And what we can use 
or not. And how they are deciding for us what our risk profile is.

**Comment by rocketraman** [comment]
> To me, in this particular case, the issue is why the KeepassXC devs are on their high horse to tell us what's good for us. And what we can use or not. And how they are deciding for us what our risk profile is.

I wouldn't go this far. The devs can get on whatever horse they want -- they are the ones putting their time and effort into building it. You aren't going to convince them to implement this feature this way. Use rational arguments (as I believe many have in this issue), and if those don't work, then fork it and build it yourself or use another tool.

Note that its pretty easy to work around the lack of this feature -- for example, this script will generate a QR code for the content you copied from KeepassXC on your Linux terminal -- I'm sure equivalents can be found for every OS:

```
xclip -selection clipboard -o text/plain | qrencode -s 6 -l H -t ANSIUTF8
```

If the devs are totally opposed to this, they can always lock this issue / limit it to contributors and/or close it and move on. The fact that they haven't tells me its less about a philosophical objection to the feature (despite the posts above) and more about just the time and effort to build something they don't need themselves, or think is very secure. If is issue is now closed and locked, we'll know otherwise :-).

**Comment by phoerious** [comment]
My barcode scanner app had a hidden history feature logging all the stuff I had ever scanned without me really realising. Fortunately, I never use it for anything sensitive except perhaps temporary tokens, which expire after use. I doubt average Joe will ever care to check their scanner for a history feature, so over time they build a silent unencrypted copy of their password database on their phone. If you are an expert user how is aware of all the implications, there are plenty of other ways to get your password to your phone like the one suggested above.

Unless we build support for encrypted QR codes via KeePass app on the phone, this will not be implemented.

**Comment by hcharbonnier** [comment]
I think maybe just allow to export data as QR Code is not the right
solution, but there is definitely a need to easily transfer data to a
smartphone.
Smartphones are rarely up to date, full of vulnerabilities and infected
apps. So opening a Keepass database on a smartphone is definitly not a
secured solution.

Maybe a solution could be to allow exporting encrypted data through a QR
code, and add a feature on the android app to decrypt this data

Le mar. 9 févr. 2021 à 11:05, Janek Bevendorff <notifications@github.com> a
écrit :

**Comment by phoerious** [comment]
> Maybe a solution could be to allow exporting encrypted data through a QR
code, and add a feature on the android app to decrypt this data

That is what I was referring to. But that would need support in the mobile apps.

**Comment by itc-ger** [comment]
> As a more general example: My barcode scanner app had a hidden history feature logging all the stuff I had ever scanned without me really realising. [..]
> 
> Unless we build support for encrypted QR codes via KeePass app on the phone, this will not be implemented.

Because there are _some_ Smartphone Apps with hidden history features, you deny to implement that feature at all? And ignoring the fact that there are many others with "disable history" options build in?

Please think about this:  Isn't it the same if users are allowed to copy passwords from KeePassXC to the clipboard?
From my point of view it's exactly the same. KeepassXC cannot guarantee that the user has installed another clipboard managing software, which creates a history. 

Therefore I would be happy if you would reconsider your decision.
